### PR TITLE
FIX all settings / restrictions are now queried correctly

### PR DESCRIPTION
**Pull request by estafilarakis**
@plentymarkets/everyone-read 
@dorin-ciceu

### CODE REVIEW

**Review by dorin-ciceu — APPROVED** [review]

File: src/Methods/InvoicePaymentMethod.php [hunk 1]
```diff
@@ -67,86 +71,36 @@ public function __construct(
...
-        {
-            return false;
-        }
-
-        if(!in_array($this->checkout->getShippingCountryId(), $this->settings->getShippingCountries()))
-        {
-            return false;
+        $isGuest = !($this->accountService->getIsAccountLoggedIn() && $basket->customerId > 0);
+        $contact = null;
+        if(!$isGuest && $basket->customerId > 0) {
```
Comment (CTohme) [inline src/Methods/InvoicePaymentMethod.php]: Why does the If-Statement include `$basket->customerId > 0`, that bool-value is already included in the variable `isGuest`?

File: src/Methods/InvoicePaymentMethod.php [hunk 2]
```diff
@@ -67,86 +71,36 @@ public function __construct(
...
+            try {
+                $contact = $contactRepo->findContactById($basket->customerId, ['orderSummary']);
+            } catch(\Exception $ex) {}
         }
-
-        return true;
+        $amount = $amount = pluginApp(OrderAmount::class);
```
Comment (CTohme) [inline src/Methods/InvoicePaymentMethod.php]: I think this may have been a copy & paste thing here? Or why did you double the allocation of the variable `$amount`?

File: src/Methods/InvoicePaymentMethod.php [hunk 3]
```diff
@@ -262,37 +207,63 @@ public function getDescription():string
...
-                /**
-                 * Check the maximum amount
-                 */
-                if( $maxAmount > 0.00 && $maxAmount < $amount->invoiceTotal) {
-                    return false;
-                }
-            } catch(\Exception $e) {}
+                return $helper->respectsAllLimitations(
+                    pluginApp(SettingsHelper::class, [$this->settings, $order->plentyId]),
+                    $contact === null,
```
Comment (CTohme) [inline src/Methods/InvoicePaymentMethod.php]: Isn't there a second attribute to check, wheither or not it deals with a guest order. The Contact-Model contains the following Method:
```php
	/**
	 * Accessor for "access type" contact option
	 *
	 * @internal
	 * @return string 1 if access type is guest, otherwise empty string
	 */
	public function getSingleAccessAttribute()
	{
		$option = $this->findOption(ContactOption::TYPE_ACCESS, ContactOption::SUBTYPE_GUEST);

		if ($option != null)
		{
			return $option->value;
		}
		return "";
	}
```
This seems to me, that an additional should be made in order to verify if an order has been placed by a guest, like so:
```php
$isGuest = !($contact === null || $contact->getSingleAccess == 1);
``` 
Or am I wrong here?

File: src/Services/InvoiceLimitationsService.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+     * @param int            $billingAddressId  The ID of the billing address
+     * @param int            $deliveryAddressId The ID of the delivery address
+     * @param Contact|null   $contact           The contact instance, if customer is not a quest
+     * 
+     * @return bool
+     */
+    public function respectsAllLimitations(
+        SettingsHelper $settingsHelper,
+        bool $isGuest,
+        OrderAmount $amount,
```
Comment (CTohme) [inline src/Services/InvoiceLimitationsService.php]: Why did you add an OrderAmount Object as parameter here? Especially since where it is used, you are "faking" this OrderAmount-Object with only setting the total amount and the currency?

Why can't you just use two parameters, which makes it more explicite to read, seeing that both a float and string are given here?

File: src/Helper/SettingsHelper.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    
+    /**
+     * Get the minimum amount for using the payment method.
+     * 
+     * @return float
+     */
+    public function minimumOrderCount(): int
+    {
+        return (int)$this->getSettingIntValue('quorumOrders');
+    }
```
Comment (kevinfink96) [inline src/Helper/SettingsHelper.php]: 1. The annotation says return `float`
2. Is the int cast really necessary, because the method `getSettingIntValue` should always return an integer and there is also a cast in it, same question for the method with float.

File: src/Services/InvoiceLimitationsService.php [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                return false;
+            }
+        }
+        
+        if($amount->invoiceTotal > 0.0) {
+            /** @var CurrencyExchangeRepositoryContract $currencyService */
+            $currencyService = pluginApp(CurrencyExchangeRepositoryContract::class);
+            
+            //  Fifth: Check minimum amount
+            $minAmount = (float)$currencyService->convertFromDefaultCurrency($amount->currency, $settingsHelper->minimumAmount());
```
Comment (CTohme) [inline src/Services/InvoiceLimitationsService.php]: Doesn't it make sense to first of all verify if `$currency` is the default currency or not?
-> The currency conversion has to be made if `$currency` is not the default currency
maybe like so:
```php
if($currencyService->isDefaultCurrency($currency))
{
$minAmount = (float)$currencyService->convertFromDefaultCurrency($amount->currency, $settingsHelper->minimumAmount());
$maxAmount = (float)$currencyService->convertFromDefaultCurrency($amount->currency, $settingsHelper->maximumAmount());
}
// ...
```

**Review by CTohme — COMMENTED** [review]
It was not to easy for me to read the method `public function respectsAllLimitations` since all the If-Statements first return false, that was not really intuitiv to me, which made it tough.
But too me it does look good. 

We can talk about my comments after the daily if you want.

File: src/Helper/SettingsHelper.php (hunk 5 above)
Comment (estafilarakis) [reply src/Helper/SettingsHelper.php]: 1. Copy&Paste Fehler aus `minimumAmount()`
2. Hast absolut recht.

File: src/Methods/InvoicePaymentMethod.php [hunk 7]
```diff
@@ -67,86 +70,33 @@ public function __construct(
      */
     public function isActive( BasketRepositoryContract $basketRepositoryContract):bool
     {
-
+        /** @var InvoiceLimitationsService $helper */
+        $helper = pluginApp(InvoiceLimitationsService::class);
```
Comment (j-konopka) [inline src/Methods/InvoicePaymentMethod.php]: Why does the variable called „helper“ and contains a service?

File: meta/documents/changelog_de.md [hunk 8]
```diff
@@ -1,6 +1,10 @@
 # Release Notes für Kauf auf Rechnung
 
-## 2.0.3 (2019-12-17)
+## 2.0.4 (2020-02-17)
+### Gefixt
```
Comment (j-konopka) [inline meta/documents/changelog_de.md:4]: Please change the date to the correct release date.

File: src/Methods/InvoicePaymentMethod.php [hunk 9]
```diff
@@ -262,37 +203,62 @@ public function getDescription():string
      */
     public function isSwitchableTo(int $orderId = null):bool
     {
+        /** @var InvoiceLimitationsService $helper */
+        $helper = pluginApp(InvoiceLimitationsService::class);
```
Comment (j-konopka) [inline src/Methods/InvoicePaymentMethod.php]: Why does the variable called „helper“ and contains a service?

**Review by j-konopka — APPROVED** [review]
